File: src/chapters/linear-systems/content/code/gauss-complete.jl
function gauss_complete_pivot(A, b)
    A = Matrix{Float64}(A); b = Vector{Float64}(b); n = length(b); col = collect(1:n)
    for k in 1:n
        sub = abs.(A[k:n, k:n]); idx = argmax(sub)
        i, j = idx[1] + k - 1, idx[2] + k - 1
        A[[k, i], :] = A[[i, k], :]; b[[k, i]] = b[[i, k]]
        A[:, [k, j]] = A[:, [j, k]]; col[[k, j]] = col[[j, k]]
        for r in k+1:n
            f = A[r, k] / A[k, k]
            A[r, k:n] .-= f .* A[k, k:n]; b[r] -= f * b[k]
        end
    end
    y = zeros(n)
    for i in n:-1:1
        y[i] = (b[i] - sum(A[i, j] * y[j] for j in i+1:n; init = 0.0)) / A[i, i]
    end
    x = zeros(n); x[col] = y
    return x
end
A = [2.0 1 -1; -3 -1 2; -2 1 2]; b = [8.0, -11, -3]
println(gauss_complete_pivot(A, b))

File: src/chapters/linear-systems/content/code/gauss-complete.rs
// Gaussian elimination with complete (row + column) pivoting.
fn gauss_complete_pivot(a: &[Vec<f64>], b: &[f64]) -> Vec<f64> {
    let n = b.len();
    let mut m: Vec<Vec<f64>> = a.iter().cloned().collect();
    let mut r = b.to_vec();
    let mut col: Vec<usize> = (0..n).collect();
    for k in 0..n {
        let (mut pi, mut pj) = (k, k);
        for i in k..n { for j in k..n { if m[i][j].abs() > m[pi][pj].abs() { pi = i; pj = j; } } }
        m.swap(k, pi); r.swap(k, pi);
        for row in m.iter_mut() { row.swap(k, pj); }
        col.swap(k, pj);
        for i in k + 1..n {
            let f = m[i][k] / m[k][k];
            for j in k..n { m[i][j] -= f * m[k][j]; }
            r[i] -= f * r[k];
        }
    }
    let mut y = vec![0.0; n];
    for i in (0..n).rev() {
        let mut s = r[i];
        for j in i + 1..n { s -= m[i][j] * y[j]; }
        y[i] = s / m[i][i];
    }
    let mut x = vec![0.0; n];
    for i in 0..n { x[col[i]] = y[i]; }
    x
}
fn main() {
    let a = vec![vec![2.0, 1.0, -1.0], vec![-3.0, -1.0, 2.0], vec![-2.0, 1.0, 2.0]];
    let b = vec![8.0, -11.0, -3.0];
    println!("{:?}", gauss_complete_pivot(&a, &b));
}
